tls/ecp_curves.c

```diff
@@ -976,203 +735,87 @@ static int ecp_mod_p384(ttls_mpi *N)
+#define P521_WIDTH	  (521 / BIL + 1)
 
-/* Bits to keep in the most significant ttls_mpi_uint */
+/* Bits to keep in the most significant unsigned long */
 #define P521_MASK	   0x01FF
 
 /*
- * Fast quasi-reduction modulo p521 (FIPS 186-3 D.2.5)
- * Write N as A1 + 2^521 A0, return A0 + A1
+ * Fast quasi-reduction modulo p521 (FIPS 186-3 D.2.5).
+ * Write N as A1 + 2^521 A0, return A0 + A1.
  */
-static int ecp_mod_p521(ttls_mpi *N)
+static int
+ecp_mod_p521(TlsMpi *N)
 {
-	int ret;
-	size_t i;
-	ttls_mpi M;
-	ttls_mpi_uint Mp[P521_WIDTH + 1];
-	/* Worst case for the size of M is when ttls_mpi_uint is 16 bits:
-	 * we need to hold bits 513 to 1056, which is 34 limbs, that is
-	 * P521_WIDTH + 1. Otherwise P521_WIDTH is enough. */
-
-	if (N->n < P521_WIDTH)
+	int r;
+	TlsMpi M;
+	unsigned long Mp[P521_WIDTH + 1];
+
+	if (N->used < P521_WIDTH)
 		return 0;
 
 	/* M = A1 */
 	M.s = 1;
-	M.n = N->n - (P521_WIDTH - 1);
-	if (M.n > P521_WIDTH + 1)
-		M.n = P521_WIDTH + 1;
+	M.limbs = P521_WIDTH + 1;
+	M.used = N->used - (P521_WIDTH - 1);
+	if (M.used > P521_WIDTH + 1)
+		M.used = P521_WIDTH + 1;
 	M.p = Mp;
-	memcpy(Mp, N->p + P521_WIDTH - 1, M.n * sizeof(ttls_mpi_uint));
-	TTLS_MPI_CHK(ttls_mpi_shift_r(&M, 521 % (8 * sizeof(ttls_mpi_uint))));
+
+	memcpy(Mp, N->p + P521_WIDTH - 1, M.used * CIL);
+	if ((r = ttls_mpi_shift_r(&M, 521 % BIL)))
+		return r;
 
 	/* N = A0 */
 	N->p[P521_WIDTH - 1] &= P521_MASK;
-	for (i = P521_WIDTH; i < N->n; i++)
-		N->p[i] = 0;
+	N->used = P521_WIDTH;
 
 	/* N = A0 + A1 */
-	TTLS_MPI_CHK(ttls_mpi_add_abs(N, N, &M));
-
-cleanup:
-	return ret;
+	return ttls_mpi_add_abs(N, N, &M);
 }
 
 #undef P521_WIDTH
 #undef P521_MASK
 
-#endif /* TTLS_ECP_NIST_OPTIM */
-
-/* Size of p255 in terms of ttls_mpi_uint */
-#define P255_WIDTH	  (255 / 8 / sizeof(ttls_mpi_uint) + 1)
+/* Size of p255 in terms of unsigned long */
+#define P255_WIDTH	  (255 / 8 / sizeof(unsigned long) + 1)
 
-/*
- * Fast quasi-reduction modulo p255 = 2^255 - 19
- * Write N as A0 + 2^255 A1, return A0 + 19 * A1
+/**
+ * Fast quasi-reduction modulo p255 = 2^255 - 19.
+ * Write N as A0 + 2^255 A1, return A0 + 19 * A1.
  */
-static int ecp_mod_p255(ttls_mpi *N)
+static int
+ecp_mod_p255(TlsMpi *N)
 {
-	int ret;
-	size_t i;
-	ttls_mpi M;
-	ttls_mpi_uint Mp[P255_WIDTH + 2];
+	int r;
+	size_t n;
+	TlsMpi M;
+	unsigned long Mp[P255_WIDTH + 2];
 
-	if (N->n < P255_WIDTH)
+	if (N->used < P255_WIDTH)
 		return 0;
 
 	/* M = A1 */
 	M.s = 1;
-	M.n = N->n - (P255_WIDTH - 1);
-	if (M.n > P255_WIDTH + 1)
-		M.n = P255_WIDTH + 1;
+	M.limbs = P255_WIDTH + 2;
+	M.used = N->used - (P255_WIDTH - 1);
+	if (M.used > P255_WIDTH + 1)
+		M.used = P255_WIDTH + 1;
 	M.p = Mp;
-	memset(Mp, 0, sizeof Mp);
-	memcpy(Mp, N->p + P255_WIDTH - 1, M.n * sizeof(ttls_mpi_uint));
-	TTLS_MPI_CHK(ttls_mpi_shift_r(&M, 255 % (8 * sizeof(ttls_mpi_uint))));
-	M.n++; /* Make room for multiplication by 19 */
+	n = M.used * CIL;
+	memcpy(Mp, N->p + P255_WIDTH - 1, n);
+	memset(Mp + n, 0, sizeof(Mp) - n);
```

Comment:
`n` is number of bytes, but `Mp` is coerced to `unsigned long *`, so we may write beyond the `Mp` here.